From 821002669b591ba0f31dfd9891bfd7617982b318 Mon Sep 17 00:00:00 2001
From: Liu Ying <victor.liu@nxp.com>
Date: Tue, 25 Jul 2017 15:25:28 +0800
Subject: [PATCH 2205/5242] MLK-16075-10 gpu: imx: dpu: common: Add helpers to
 get plane w/h of format

commit  c8667523aae3b0658327dc877ac5a269d7f05303 from
https://source.codeaurora.org/external/imx/linux-imx.git

This patch adds helpers dpu_format_plane_width/height() to get plane
width or height of pixel formats which are supported by the current
dpu base driver.

Signed-off-by: Liu Ying <victor.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dpu/dpu-common.c |   22 ++++++++++++++++++++++
 drivers/gpu/imx/dpu/dpu-prv.h    |    2 ++
 2 files changed, 24 insertions(+)

diff --git a/drivers/gpu/imx/dpu/dpu-common.c b/drivers/gpu/imx/dpu/dpu-common.c
index 2f280b7..1622fdb 100644
--- a/drivers/gpu/imx/dpu/dpu-common.c
+++ b/drivers/gpu/imx/dpu/dpu-common.c
@@ -590,6 +590,28 @@ int dpu_format_num_planes(u32 format)
 	}
 }
 
+int dpu_format_plane_width(int width, u32 format, int plane)
+{
+	if (plane >= dpu_format_num_planes(format))
+		return 0;
+
+	if (plane == 0)
+		return width;
+
+	return width / dpu_format_horz_chroma_subsampling(format);
+}
+
+int dpu_format_plane_height(int height, u32 format, int plane)
+{
+	if (plane >= dpu_format_num_planes(format))
+		return 0;
+
+	if (plane == 0)
+		return height;
+
+	return height / dpu_format_vert_chroma_subsampling(format);
+}
+
 #define DPU_UNIT_INIT(dpu, base, unit, name, id, pec_ofs, ofs)		\
 {									\
 	int ret;							\
diff --git a/drivers/gpu/imx/dpu/dpu-prv.h b/drivers/gpu/imx/dpu/dpu-prv.h
index f8e6e7f..04f0e89 100644
--- a/drivers/gpu/imx/dpu/dpu-prv.h
+++ b/drivers/gpu/imx/dpu/dpu-prv.h
@@ -229,6 +229,8 @@ struct dpu_soc {
 int dpu_format_horz_chroma_subsampling(u32 format);
 int dpu_format_vert_chroma_subsampling(u32 format);
 int dpu_format_num_planes(u32 format);
+int dpu_format_plane_width(int width, u32 format, int plane);
+int dpu_format_plane_height(int height, u32 format, int plane);
 
 #define DECLARE_DPU_UNIT_INIT_FUNC(block)			\
 int dpu_##block##_init(struct dpu_soc *dpu, unsigned int id,	\
-- 
1.7.9.5

